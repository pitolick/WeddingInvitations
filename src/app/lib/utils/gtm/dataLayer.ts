/**
 * dataLayerPushの処理を定義する
 * @see https://qiita.com/cheez921/items/a9e8d257684098db55c3
 */
type PageViewEvent = {
  event: 'NextRoute';
  pagePath: string;
};

type InViewEvent = {
  event: 'NextInView';
  label: string;
};

type ClickEvent = {
  event: 'NextClick';
  label: string;
};

// windowを拡張
declare global {
  interface Window {
    dataLayer: Record<string, unknown>[];
  }
}

export type DataLayerType = PageViewEvent | InViewEvent | ClickEvent;

export const pushDataLayer = (data: DataLayerType): void => {
  if (typeof window === 'undefined') return;
  window.dataLayer = window.dataLayer || [];
  window.dataLayer.push(data);
};
