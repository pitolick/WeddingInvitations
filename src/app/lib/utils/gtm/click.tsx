/**
 * クリックイベントをトラッキングしてGTMに送信
 * @see https://qiita.com/cheez921/items/a9e8d257684098db55c3
 */
import React, { FC, ReactNode, useCallback } from 'react';
import { pushDataLayer } from './dataLayer';

type Props = {
  children: ReactNode;
  label: string;
};

export const TrackOnClick: FC<Props> = ({ children, label }) => {
  const handleClick = useCallback(() => {
    pushDataLayer({
      event: 'NextClick',
      label,
    });
  }, [label]);

  return <div onClick={handleClick}>{children}</div>;
};
