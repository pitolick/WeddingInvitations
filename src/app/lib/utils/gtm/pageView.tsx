/**
 * ページビューをトラッキングしてGTMに送信
 * @see https://qiita.com/cheez921/items/a9e8d257684098db55c3
 * @todo テストコード作る（return null のFCをどのように取得し動作確認したらいいか不明のため断念）
 */
'use client';
import { FC, useEffect, useCallback } from 'react';
import { useSearchParams, usePathname } from 'next/navigation';
import { pushDataLayer } from './dataLayer';

export const TrackPageView: FC = () => {
  const searchParams = useSearchParams();
  const pathname = usePathname();

  // NextRouteイベントを送信
  const pushPageView = useCallback((url: string) => {
    pushDataLayer({
      event: 'NextRoute',
      pagePath: url,
    });
  }, []);

  // 初期ロード時の処理
  useEffect(() => {
    // routeChangeCompleteが返すurlと同じ形になるよう整形
    const url =
      `${pathname}${searchParams.toString() && '?' + searchParams.toString()}`.replace(
        /\/$|\/\?/,
        matched => {
          return matched === '/?' ? '?' : '';
        }
      );
    pushPageView(url);
  }, [pushPageView, pathname, searchParams]);

  return null;
};
