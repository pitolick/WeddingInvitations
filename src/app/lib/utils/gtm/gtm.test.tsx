import { render } from '@testing-library/react';

// 環境変数をモック
const mockGTM_ID = 'GTM-TEST123';
process.env.NEXT_PUBLIC_GTM_ID = mockGTM_ID;

// モックの設定
jest.mock('next/script', () => ({
  __esModule: true,
  default: jest.fn(() => <></>),
}));

jest.mock('./pageView', () => ({
  __esModule: true,
  TrackPageView: jest.fn(() => <></>),
}));

import { GoogleTagManagerHead, GoogleTagManagerBody } from './gtm';

describe('GoogleTagManagerHead', () => {
  it('コンポーネントが正しくレンダリングされること', () => {
    const { container } = render(<GoogleTagManagerHead />);
    expect(container).toMatchSnapshot();
  });
});

describe('GoogleTagManagerBody', () => {
  it('コンポーネントが正しくレンダリングされること', () => {
    const { container } = render(<GoogleTagManagerBody />);
    expect(container).toMatchSnapshot();
  });
});
