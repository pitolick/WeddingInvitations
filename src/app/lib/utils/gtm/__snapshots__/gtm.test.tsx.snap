// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleTagManagerBody コンポーネントが正しくレンダリングされること 1`] = `
<div>
  <noscript>
    &lt;iframe src="https://www.googletagmanager.com/ns.html?id=GTM-TEST123" height="0" width="0" style="display:none;visibility:hidden" /&gt;
  </noscript>
</div>
`;

exports[`GoogleTagManagerHead コンポーネントが正しくレンダリングされること 1`] = `<div />`;
